### PR TITLE
Add Forgot password flow

### LINKED ISSUE #586: Add ForgotPasswordTapped action and wire up the button
## Description
Add ForgotPasswordTapped action and wire up the button

## Definition of done

- Action is added and reduced
- Tests are added
- Button is added to the login view
- Action is emitted  
- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #588: Forgot password - add EmailEntered action
## Description
Add the action for handling email input on the forgot password screen. Wire it up to the UI if it’s already available.
## Definition of done
- Action is added and reduced
- It’s wired with the UI if the UI is there
- Tests are added
- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #589: Forgot password - add CloseButtonTapped action
## Description
Add the action handling closing the forgot password view. Wire it up to the UI if it’s there.
## Definition of done
- Action is added and reduced
- It’s wired with the UI if the UI is there
- Tests are added
- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #590: Forgot password - add PasswordResetEmailSent and PasswordResetEmailFailed action
## Description
Add the relevant actions. Display an appropriate toast when email is successfully sent. Set an error if the reset email couldn’t be sent.
## Definition of done
- Actions are added and reduced
- They are wired to the UI if the UI is there
- Tests are added

- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #591: Forgot password - add SendEmailButtonTapped action, SendPasswordResetEmailEffect and the API interaction
## Description
Add the action and effect. Wire the action up with the UI if it’s there. See the Xamarin implementation for request/response spec.
## Definition of done
- Action is added and reduced
- Effect is added
- API is integrated
- Tests are added

- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #674: System Windows are popping the backstack somehow
## Description
Any kind of system dialog (e.g.: Asking for user permissions/opening the password manager) is navigating back to the previous app screen

## Definition of done
- [ ] The user is no longer taken to the previous screen when accepting permissions
- [ ] The user is no longer taken to the previous screen when using a password manager

### PR DESCRIPTION
**Pull request by heytherewill**
### Summary
This adds the flow needed for the password reset thingy
This also fixes the backstack shenanigans because it's IMPOSSIBLE to test this without using a password manager 😂 

### Technical considerations
The exception handling is taken from the old apps and is based [on this](https://github.com/toggl/operations/blob/master/infra/http_statuses.md). It doesn't handle SSO exceptions; these should be handled when we actually implement SSO

### Relationships
Closes #586 
Closes #588
Closes #589 
Closes #590
Closes #591
Closes #674 

### Pull Request checklist
- [X] The code does not hinder the app's performance
- [X] The code does not introduce regressions
- [X] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests
- [ ] One of our [live or file custom templates](https://github.com/toggl/android/tree/master/.idea/templates) was used in this PR

### Pointers
Going at it commit by commit will prove very helpful. It's a lot of files touched, but the changes aren't hard to follow if you look at it commit by commit 

### Review template
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones
- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Only myself

### CODE REVIEW

File: api/src/main/java/com/toggl/api/authentication/RetrofitAuthenticationApiClient.kt [hunk 1]
```diff
@@ -13,4 +14,10 @@ internal class RetrofitAuthenticationApiClient(
         val authHeader = email.basicAuthenticationWithPassword(password)
         return authenticationApi.login(authHeader)
     }
+
+    override suspend fun resetPassword(email: Email.Valid): String {
+        val body = ResetPasswordBody(email.toString())
+        val result = authenticationApi.resetPassword(body)
+        return result.trim('"')
```
Comment (daividssilverio) [inline api/src/main/java/com/toggl/api/authentication/RetrofitAuthenticationApiClient.kt:21]: what's with the trim?

File: onboarding/src/main/java/com/toggl/onboarding/passwordreset/domain/SendPasswordResetEmailEffect.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    private val email: Email.Valid
+) : Effect<PasswordResetAction> {
+    override suspend fun execute(): PasswordResetAction? = withContext(dispatcherProvider.io) {
+        try {
+            val message = authenticationApiClient.resetPassword(email)
+            PasswordResetAction.PasswordResetEmailSent(message)
+        } catch(ex: Exception) {
+
+            val failure = Failure(ex, when (ex) {
+                is BadRequestException -> emailDoesNotExistMessage
```
Comment (daividssilverio) [inline onboarding/src/main/java/com/toggl/onboarding/passwordreset/domain/SendPasswordResetEmailEffect.kt:29]: Not sure if we should send this error message 🤔 Isn't it a security exception?
Shouldn't we always show something like "if your e-mail exists in our database, you'll receive a link with the reset instructions" regardless of the result of the request? (We have to handle offline/server errors though).

File: api/src/main/java/com/toggl/api/authentication/RetrofitAuthenticationApiClient.kt (hunk 1 above)
Comment (heytherewill) [reply api/src/main/java/com/toggl/api/authentication/RetrofitAuthenticationApiClient.kt:21]: Copied from Xamarin. Past William/Simon probably knew better and I trust their wisdom

File: onboarding/src/main/java/com/toggl/onboarding/login/ui/LoginFragment.kt [hunk 3]
```diff
@@ -26,22 +30,52 @@ class LoginFragment : Fragment(R.layout.fragment_login) {
...
+
+        store.state
+            .map { if (it.resetPasswordResult is Loadable.Loaded) it.resetPasswordResult.value else "" }
+            .filter { it.isNotBlank() }
+            .onEach { Toast.makeText(requireContext(), it, Toast.LENGTH_LONG).show() }
+            .launchIn(lifecycleScope)
+    }
+
+    override fun onDestroyView() {
+        email.removeTextChangedListener(emailWatcher)
```
Comment (daividssilverio) [inline onboarding/src/main/java/com/toggl/onboarding/login/ui/LoginFragment.kt:76]: If you are considering that the views might be null `onResume`, please consider them as nullable here as well.
The doc says the view can be null.
```
/**
     * Called when the view previously created by {@link #onCreateView} has
     * been detached from the fragment.  The next time the fragment needs
     * to be displayed, a new view will be created.  This is called
     * after {@link #onStop()} and before {@link #onDestroy()}.  It is called
     * <em>regardless</em> of whether {@link #onCreateView} returned a
     * non-null view.  Internally it is called after the view's state has
     * been saved but before it has been removed from its parent.
     */
```

File: onboarding/src/main/java/com/toggl/onboarding/login/ui/LoginFragment.kt [hunk 4]
```diff
@@ -25,29 +30,52 @@ class LoginFragment : Fragment(R.layout.fragment_login) {
...
+            .onEach { password?.setSafeText(it.toString()) }
+            .launchIn(lifecycleScope)
 
         store.state
             .map { it.email is Email.Valid && it.password is Password.Valid && it.user !is Loadable.Loading }
             .distinctUntilChanged()
             .onEach { login_button.isEnabled = it }
             .launchIn(lifecycleScope)
+
+        store.state
```
Comment (daividssilverio) [inline onboarding/src/main/java/com/toggl/onboarding/login/ui/LoginFragment.kt:68]: This toast is being fired multiple times. Maybe throw in a `.distinctUntilChanged()`?

Comment (daividssilverio) [reply onboarding/src/main/java/com/toggl/onboarding/login/ui/LoginFragment.kt:68]: It also fires every time you change the state in any other way after `resetPasswordResult` has been set.

File: onboarding/src/main/java/com/toggl/onboarding/passwordreset/ui/PasswordResetFragment.kt [hunk 5]
```diff
@@ -1,11 +1,38 @@
...
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
 
 @AndroidEntryPoint
 class PasswordResetFragment : Fragment(R.layout.fragment_passwordreset) {
     private val store: PasswordResetStoreViewModel by viewModels()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
```
Comment (daividssilverio) [inline onboarding/src/main/java/com/toggl/onboarding/passwordreset/ui/PasswordResetFragment.kt:26]: Shouldn't you at least show a toast when an error happens in the request?

**Comment by heytherewill** [comment]
@daividssilverio The Up navigation is not doable because there's no toolbar. We are in dire need of designs atm, so let's put this on hold for now, since the user can always use the physical back button

**Comment by heytherewill** [comment]
@pdrj Let's fix animations in another PR, since we don't know if this will remain full screen or become a modal

File: models/src/main/java/com/toggl/models/validation/Email.kt [hunk 6]
```diff
@@ -8,8 +8,11 @@ sealed class Email(val email: String) {
...
+            private val emailRegex = Regex(pattern, RegexOption.IGNORE_CASE)
+
             fun from(email: String) =
-                if (email.isBlank()) Invalid(email)
+                if (email.isBlank() || !emailRegex.matches(email)) Invalid(email)
```
Comment (daividssilverio) [inline models/src/main/java/com/toggl/models/validation/Email.kt:15]: why not `android.util.Patterns.EMAIL_ADDRESS` ?

Comment (heytherewill) [reply models/src/main/java/com/toggl/models/validation/Email.kt:15]: This model is not in an Android module

File: api/src/main/java/com/toggl/api/clients/ErrorHandlingProxyClient.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import com.toggl.models.validation.Email
+import com.toggl.models.validation.Password
+import retrofit2.HttpException
+import java.lang.Exception
+import java.net.UnknownHostException
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+internal class ErrorHandlingProxyClient @Inject constructor(
```
Comment (daividssilverio) [inline api/src/main/java/com/toggl/api/clients/ErrorHandlingProxyClient.kt:21]: Every api call will have to go through the proxy from now on?

Comment (heytherewill) [reply api/src/main/java/com/toggl/api/clients/ErrorHandlingProxyClient.kt:21]: Yes. Not sure how ideal that is, but the interceptor just rethrows and gives us no chance to handle it 🤔

Comment (heytherewill) [reply api/src/main/java/com/toggl/api/clients/ErrorHandlingProxyClient.kt:21]: We can always revisit that later if it becomes a hassle

File: models/src/main/java/com/toggl/models/validation/Email.kt (hunk 6 above)
Comment (daividssilverio) [reply models/src/main/java/com/toggl/models/validation/Email.kt:15]: This is mildly annoying, but 🤷‍♂️ .
